runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/StreamingEngineClient.java

```diff
@@ -0,0 +1,444 @@
+
+    return ImmutableList.<Long>builder()
+        .add(getWorkerMetadataThrottleTimer.getAndResetThrottleTime())
+        .addAll(keyedWorkStreamThrottleTimes)
+        .build();
+  }
+
+  /** Starts {@link GetWorkerMetadataStream}. */
+  @SuppressWarnings({
+    "ReturnValueIgnored", // starts the stream, this value is memoized.
+  })
+  private void startGetWorkerMetadataStream() {
+    getWorkerMetadataStream.get();
+  }
+
+  private synchronized ImmutableMap<Endpoint, WindmillConnection> createNewWindmillConnections(
+      ImmutableSet<Endpoint> newWindmillEndpoints) {
+    ImmutableMap<Endpoint, WindmillConnection> currentConnections =
+        connections.get().windmillConnections();
+    return newWindmillEndpoints.stream()
+        .collect(
+            toImmutableMap(
+                Function.identity(),
+                // Reuse existing stubs if they exist.
+                endpoint ->
+                    currentConnections.getOrDefault(
+                        endpoint, WindmillConnection.from(endpoint, this::createWindmillStub))));
+  }
+
+  private synchronized ImmutableMap<WindmillConnection, WindmillStreamSender>
+      closeStaleStreamsAndCreateNewStreams(
+          ImmutableSet<WindmillConnection> newWindmillConnections) {
+    ImmutableMap<WindmillConnection, WindmillStreamSender> currentStreams =
+        connections.get().windmillStreams();
+
+    // Close the streams that are no longer valid.
+    currentStreams.entrySet().stream()
+        .filter(
+            connectionAndStream -> !newWindmillConnections.contains(connectionAndStream.getKey()))
+        .map(Map.Entry::getValue)
+        .forEach(WindmillStreamSender::closeAllStreams);
+
+    return newWindmillConnections.stream()
+        .collect(
+            toImmutableMap(
+                Function.identity(),
+                newConnection ->
+                    Optional.ofNullable(currentStreams.get(newConnection))
+                        .orElseGet(() -> createWindmillStreamSenderWithNoBudget(newConnection))));
+  }
+
+  private ImmutableMap<Endpoint, Supplier<GetDataStream>> createNewGlobalDataStreams(
+      ImmutableSet<Endpoint> newGlobalDataEndpoints) {
+    ImmutableMap<Endpoint, Supplier<GetDataStream>> currentGlobalDataStreams =
+        connections.get().globalDataStreams();
+
+    return newGlobalDataEndpoints.stream()
+        .collect(
+            toImmutableMap(
+                Function.identity(),
+                endpoint -> existingOrNewGetDataStreamFor(endpoint, currentGlobalDataStreams)));
+  }
+
+  private Supplier<GetDataStream> existingOrNewGetDataStreamFor(
+      Endpoint endpoint, ImmutableMap<Endpoint, Supplier<GetDataStream>> currentGlobalDataStreams) {
+    return Preconditions.checkNotNull(
+        currentGlobalDataStreams.getOrDefault(
+            endpoint,
+            () ->
+                streamFactory.createGetDataStream(
+                    WindmillConnection.from(endpoint, this::createWindmillStub).stub(),
```

Comment:
If we already have a connection for this endpoint due to windmill endpoints, use that stub.

Reply:
done.